File: ios/Sources/FarmerChatSDK/Models/UI/ChatState.swift
import Foundation

// MARK: - ChatState

/// Immutable snapshot of the entire chat screen state.
public struct ChatState {
    /// All messages currently visible in the thread (user + AI + loading).
    public var messages: [ChatMessage] = []

    /// Suggested follow-up question texts shown in the bottom chip bar.
    public var suggestedQuestions: [String]?

    /// IDs corresponding to `suggestedQuestions` (same index mapping).
    public var suggestedQuestionIds: [String]?

    /// True while an API call is in-flight.
    public var isLoading: Bool = false

    /// Non-nil when an error should be shown inline in the thread.
    public var errorMessage: String?

    /// ID of the AI message that failed (used for retry UI).
    public var failedMessageId: String?

    /// True while an audio synthesis request is in-flight.
    public var isLoadingSynthesiseAudio: Bool = false

    /// URL of the latest synthesised audio, ready for playback.
    public var audioPlaybackUrl: URL?

    /// True while audio is actively playing.
    public var isAudioPlaying: Bool = false

    /// The next page number for paginated history loading, nil when all loaded.
    public var historyNextPage: Int?

    /// True after the first page of chat history has been fetched.
    public var isInitialHistoryLoaded: Bool = false

    /// The active conversation ID (set after the first message is sent).
    public var conversationId: String?

    public init() {}
}

File: android/farmerchat-sdk/src/main/java/com/farmerchat/sdk/ui/chat/udf/ChatState.kt
package com.farmerchat.sdk.ui.chat.udf

import com.farmerchat.sdk.base.UiState
import com.farmerchat.sdk.ui.chat.model.ChatMessage

data class ChatState(
    val messages: List<ChatMessage> = emptyList(),
    val suggestedQuestions: List<String>? = null,
    val suggestedQuestionIds: List<String>? = null,
    val chatResponseState: UiState<String> = UiState.Idle,
    val errorMessage: String? = null,
    val failedMessageId: String? = null,
    val isLoading: Boolean = false,
    val isLoadingSynthesiseAudio: Boolean = false,
    val audioPlaybackUrl: String? = null,
    val isAudioPlaying: Boolean = false,
    val historyNextPage: Int? = null,
    val isInitialHistoryLoaded: Boolean = false,
    val readFullAdviceRequestedForMessageId: String? = null,
    val currentConversationId: String? = null
)
